include/deal.II/base/utilities.h

```diff
@@ -1217,31 +1214,21 @@ namespace Utilities
         // use buffer as the target of a compressing
         // stream into which we serialize the current object
         const std::size_t previous_size = dest_buffer.size();
+        {
+          boost::iostreams::filtering_ostreambuf fosb;
 #ifdef DEAL_II_WITH_ZLIB
-        if (allow_compression)
-          {
-            boost::iostreams::filtering_ostream out;
-            out.push(
-              boost::iostreams::gzip_compressor(boost::iostreams::gzip_params(
-                boost::iostreams::gzip::default_compression)));
-            out.push(boost::iostreams::back_inserter(dest_buffer));
-
-            boost::archive::binary_oarchive archive(out);
-            archive << object;
-            out.flush();
-          }
-        else
+          if (allow_compression)
+            fosb.push(boost::iostreams::gzip_compressor());
+#else
+          (void)allow_compression;
 #endif
-          {
-            std::ostringstream              out;
-            boost::archive::binary_oarchive archive(out);
-            archive << object;
-
-            const std::string s = out.str();
-            dest_buffer.reserve(dest_buffer.size() + s.size());
```

Comment:
Working on the benchmark test for #11364 also revealed a bug in the current implementation.

I wanted to pack several `CellId` objects into one container and reserved sufficient memory for `dest_buffer` container to do so. However, this line overwrites my carefully laid out memory management and reserves new memory for each of the 16 million `CellId` objects my test wanted to pack. This is not intentional :)

With only 250k `CellId` objects, the `pack` serialization took about `15s` in release mode on my test system, while all other serialization methods took about `0.3s`. I'll publish the results of the benchmark in the discussion of the above mentioned PR soon.

Reply:
Just for my own education, what is happening is that you've nicely reserved a very large buffer, and then this call comes along and reserves a smaller size thereby releasing memory that you need to allocate again later?

Reply:
Sorry that I did not make this more clear. Yes, that's exactly what's happening. I think for this case I reserved more than 1GB of memory for the buffer before starting the serialization operation. When saving the first `CellId` with `Utilities::pack`, the buffer reservation will be overwritten to hold less than 100 bytes, thus releasing all the carefully reserved memory. Since I wanted to save about 16mio `CellId` objects, we continue to expand the size of the buffer for every single one of these objects, which ultimately infers moving the continuous memory chunk assigned to the container multiple times.

Reply:
I wonder whether it would make sense to use `std::stack` for this instead of `std::vector`. In general, we won't know how large the final result is going to be, and it would be useful to store into an object that doesn't have to copy everything when it finds that it needs more memory. `std::stack` is such a container.